### PR TITLE
Add Distance method to Point

### PR DESCRIPTION
**Pull request by robloo**
## What does the pull request do?

Add a method that calculates the Euclidian distance between two points.

This was a helper method on the app side but seems quite useful in the framework since vector and other types already have it. Calculating distance between points seems common.

## What is the current behavior?

Distance must be calculated manually or after converting to vectors.

## What is the updated/expected behavior with this PR?

Distance can be calculated directly.

## How was the solution implemented (if it's not obvious)?

It was implemented for the two-dimensional plane following the standard Euclidian geometry formula: https://en.wikipedia.org/wiki/Euclidean_distance#Two_dimensions. An additional optimization was done to avoid Math.Pow(x, 2) usage. Instead the squares are calculated using faster multiplication.

## Checklist

- ~[ ] Added unit tests (if possible)?~
- [x] Added XML documentation to any related classes?
- ~[ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation~

## Breaking changes

None

## Obsoletions / Deprecations

None

## Fixed issues

N/A

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0036751-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Comment by Mrxx99** [comment]
Could an DistanceTo(Point) extension method also be added to point as a static method on Point is not easily discoverable.

**Comment by robloo** [comment]
I wouldn't add it as an extension method if it's useful, we should add it directly to the struct. Are there examples of DistanceTo I'm unaware of? I'm currently following (with a static method) what is done several other places.

File: src/Avalonia.Base/Point.cs [hunk 1]
```diff
@@ -164,6 +164,19 @@ public Point(double x, double y)
...
 
+        /// <summary>
+        /// Computes the Euclidean distance between the two given points.
+        /// </summary>
+        /// <param name="value1">The first point.</param>
+        /// <param name="value2">The second point.</param>
+        /// <returns>The Euclidean distance.</returns>
+        public static double Distance(Point value1, Point value2)
+        {
+            double distanceSquared = ((value2.X - value1.X) * (value2.X - value1.X)) +
```
Comment (hez2010) [inline src/Avalonia.Base/Point.cs:175]: I suggest using `Math.Pow` here.

**Comment by workgroupengineering** [comment]
> Could an DistanceTo(Point) extension method also be added to point as a static method on Point is not easily discoverable?

I agree, in other uses less memory.

```csharp

    internal struct UserStruct
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public bool Active { get; set; }
    }

    internal class UserStructWithMethod
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public bool Active { get; set; }

        public static bool IsSame(UserStructWithMethod a, UserStructWithMethod b)
        {
            return false;
        }
    }

    [RankColumn]
    [MemoryDiagnoser]
    public class StructBenchmark
    {
        List<UserStruct>? usersStructLlist;
        List<UserStructWithMethod>? usersStructLlist2;

        [Params(10, 100, 1000)]
        public int size;


        [Benchmark]
        public int UserStruct()
        {
            usersStructLlist = new List<UserStruct>(size);

            for (int i = 1; i < size; i++)
            {
                var user = new UserStruct();
                user.Id = i;
                user.Name = "User_" + i;
                user.Active = true;

                usersStructLlist.Add(user);
            }
            return usersStructLlist.Count;
        }

        [Benchmark]
        public int UserStructWithMethod()
        {
            usersStructLlist2 = new List<UserStructWithMethod>(size);

            for (int i = 1; i < size; i++)
            {
                var user = new UserStructWithMethod();
                user.Id = i;
                user.Name = "User_" + i;
                user.Active = true;

                usersStructLlist2.Add(user);
            }
            return usersStructLlist2.Count;
        }
    }

```

``` ini

BenchmarkDotNet=v0.13.5, OS=Windows 10 (10.0.19045.3086/22H2/2022Update)
Intel Core i5-10400 CPU 2.90GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK=7.0.202
  [Host]     : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2


```
|               Method | size |        Mean |     Error |    StdDev | Rank |    Gen0 |   Gen1 | Allocated |
|--------------------- |----- |------------:|----------:|----------:|-----:|--------:|-------:|----------:|
|           **UserStruct** |   **10** |    **150.8 ns** |   **0.84 ns** |   **0.79 ns** |    **1** |  **0.0918** |      **-** |     **576 B** |
| UserStructWithMethod |   10 |    195.3 ns |   3.95 ns |   7.80 ns |    2 |  0.1249 | 0.0002 |     784 B |
|           **UserStruct** |  **100** |  **2,056.0 ns** |  **41.13 ns** |  **57.66 ns** |    **3** |  **1.3542** | **0.0267** |    **8496 B** |
| UserStructWithMethod |  100 |  2,527.1 ns |  48.37 ns |  97.71 ns |    4 |  1.7319 | 0.0496 |   10864 B |
|           **UserStruct** | **1000** | **21,970.6 ns** | **290.41 ns** | **226.73 ns** |    **5** | **13.9465** | **2.6245** |   **87696 B** |
| UserStructWithMethod | 1000 | 27,714.5 ns | 545.17 ns | 708.88 ns |    6 | 17.7612 | 4.3945 |  111664 B |

File: src/Avalonia.Base/Point.cs (hunk 1 above)
Comment (robloo) [reply src/Avalonia.Base/Point.cs:175]: Why? As noted in the PR description Math.Pow() is an unnecessary overhead here. The code is like this for performance.

**Comment by robloo** [comment]
@workgroupengineering I'm shocked there is a noticeable effect on performance in both execution and memory. Static methods should not be created per instance. It is also unused. Memory increase makes no sense for. I'm tempted to say something is wrong with the compiler... this needs to be understood more.

Otherwise, from an API standpoint it seems clear to me we should follow what is done for Vector and other types.

**Comment by MrJul** [comment]
> ```cs
>     internal class UserStructWithMethod
> ```

You got a `class` here.

There should be no difference between the two (as `struct`): both have the same size (16 bytes on x64) and performance.
All allocations come from the list and strings.

**Comment by workgroupengineering** [comment]
> @workgroupengineering I'm shocked there is a noticeable effect on performance in both execution and memory. Static methods should not be created per instance. It is also unused. Memory increase makes no sense for. I'm tempted to say something is wrong with the compiler... this needs to be understood more.
> 
> Otherwise, from an API standpoint it seems clear to me we should follow what is done for Vector and other types.

è il mio errore

**Review by maxkatz6 — APPROVED** [review]

**Pull request merged**